skipper.go

```diff
@@ -1233,11 +1242,27 @@ func listenAndServeQuit(
 		}
 	}
 
+	if o.EnableHttp2Cleartext {
+		h2srv := &http2.Server{}
+		srv.Handler = h2c.NewHandler(srv.Handler, h2srv)
+
+		// Work around https://github.com/golang/go/issues/26682
+		// http2.ConfigureServer registers unexported h2srv graceful shutdown handler on srv shutdown -
+		// it calls srv.RegisterOnShutdown(h2srv.state.startGracefulShutdown).
+		// h2srv graceful shutdown handler sends GOAWAY frame to all connections and closes them after predefined delay.
+		//
+		// srv.Shutdown() runs h2srv shutdown handler in a goroutine so a special snet.ShutdownListener
+		// waits until all connections are closed.
+		http2.ConfigureServer(srv, h2srv)
+	}
+
 	log.Infof("Listen on %v", address)
 
-	l, err := listen(o, address, mtr)
-	if err != nil {
+	var listener *snet.ShutdownListener
+	if l, err := listen(o, address, mtr); err != nil {
 		return err
+	} else {
+		listener = snet.NewShutdownListener(l)
```

Comment:
This will also wait for hijacked connection when h2c is not enabled.
Need to think if there is any downside to it and maybe add a test.

Reply:
https://pkg.go.dev/net/http#Server.Shutdown
> Shutdown does not attempt to close nor wait for hijacked connections such as WebSockets. The caller of Shutdown should separately notify such long-lived connections of shutdown and wait for them to close, if desired. See RegisterOnShutdown for a way to register shutdown notification functions.

This listener could be useful by itself to track hijacked connections and maybe forcibly close and log remaining after timeout as currently skipper will be just terminated after termination grace period (assuming cluster ingress mode).

Reply:
Can you create a follow up issue for that?

Reply:
I plan to check this before we merge.